{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Micky Malvino Kusandiwinata - 2602174522 \n",
    "## Comp Bio assignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input DNA =  TTACGA\n",
      "Complement =  AATGCT\n",
      "mRNA =  AAUGCU\n",
      "Aminoacid =  Asn (N) - Ala (A)\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Translate a DNA sequence ->  mRNA (using 'U' instead of 'Thr (T) - ') -> into an aminoacid sequence (protein)\n",
    "PS: The input should be multiple of 3 (example: 3 nucleotides, 6 nucleotides, etc.)\n",
    "'''\n",
    "\n",
    "from textwrap import wrap\n",
    "\n",
    "# DNA to mRNA codon mapping\n",
    "codon_table = {\n",
    "    'AUG': 'Met (M)', 'UUU': 'Phe (F)', 'UUC': 'Phe (F)', 'UUA': 'Leu (L)', 'UUG': 'Leu (L)', \n",
    "    'CUU': 'Leu (L)', 'CUC': 'Leu (L)', 'CUA': 'Leu (L)', 'CUG': 'Leu (L)', 'AUU': 'Ile (I)', \n",
    "    'AUC': 'Ile (I)', 'AUA': 'Ile (I)', 'GUU': 'Val (V)', 'GUC': 'Val (V)', 'GUA': 'Val (V)', \n",
    "    'GUG': 'Val (V)', 'UCU': 'Ser (S)', 'UCC': 'Ser (S)', 'UCA': 'Ser (S)', 'UCG': 'Ser (S)', \n",
    "    'CCU': 'Pro (P)', 'CCC': 'Pro (P)', 'CCA': 'Pro (P)', 'CCG': 'Pro (P)', 'ACU': 'Thr (T)', \n",
    "    'ACC': 'Thr (T)', 'ACA': 'Thr (T)', 'ACG': 'Thr (T)', 'GCU': 'Ala (A)', 'GCC': 'Ala (A)', \n",
    "    'GCA': 'Ala (A)', 'GCG': 'Ala (A)', 'UAU': 'Tyr (Y)', 'UAC': 'Tyr (Y)', 'UGU': 'Cys (C)', \n",
    "    'UGC': 'Cys (C)', 'UGG': 'Trp (W)', 'CAA': 'Gln (Q)', 'CAG': 'Gln (Q)', 'AAU': 'Asn (N)', \n",
    "    'AAC': 'Asn (N)', 'GAU': 'Asp (D)', 'GAC': 'Asp (D)', 'GAA': 'Glu (E)', 'GAG': 'Glu (E)', \n",
    "    'CAU': 'His (H)', 'CAC': 'His (H)', 'AAA': 'Lys (K)', 'AAG': 'Lys (K)', 'CGU': 'Arg (R)', \n",
    "    'CGC': 'Arg (R)', 'CGA': 'Arg (R)', 'CGG': 'Arg (R)', 'AGA': 'Arg (R)', 'AGG': 'Arg (R)', \n",
    "    'GGU': 'Gly (G)', 'GGC': 'Gly (G)', 'GGA': 'Gly (G)', 'GGG': 'Gly (G)', 'UAA': 'Stop', \n",
    "    'UAG': 'Stop', 'UGA': 'Stop'\n",
    "}\n",
    "\n",
    "def translate_dna_to_protein(dna_sequence):\n",
    "    # Validate that the length is a multiple of 3\n",
    "    if len(dna_sequence) % 3 != 0:\n",
    "        raise ValueError(\"The input DNA sequence must be a multiple of 3 in length.\")\n",
    "    \n",
    "    # Convert DNA to uppercase\n",
    "    dna_sequence = dna_sequence.upper()\n",
    "\n",
    "    # DNA complement (A->T, T->A, C->G, G->C)\n",
    "    complement = dna_sequence.replace('A', 't').replace('T', 'a').replace('C', 'g').replace('G', 'c').upper()\n",
    "    \n",
    "    # Transcription: DNA to mRNA (T -> U)\n",
    "    mrna = complement.replace('T', 'U')\n",
    "    \n",
    "    # Translate mRNA to Protein\n",
    "    protein_sequence = []\n",
    "    mrna_codons = wrap(mrna, 3)\n",
    "    \n",
    "    for codon in mrna_codons:\n",
    "        amino_acid = codon_table.get(codon, 'X')  # 'X' is for unknown codon\n",
    "        if amino_acid == 'Stop':\n",
    "            break\n",
    "        protein_sequence.append(amino_acid)\n",
    "    \n",
    "    # Format the protein sequence\n",
    "    protein_sequence_formatted = ' - '.join(protein_sequence)\n",
    "\n",
    "    # Output the results in the desired format\n",
    "    print(\"Input DNA = \", dna_sequence)\n",
    "    print(\"Complement = \", complement)\n",
    "    print(\"mRNA = \", mrna)\n",
    "    print(\"Aminoacid = \", protein_sequence_formatted)\n",
    "\n",
    "# Example usage\n",
    "dna_sequence = \"TTACGA\"  # Example input, should be a multiple of 3\n",
    "translate_dna_to_protein(dna_sequence)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Input Aminoacid = WYW\n",
      "\n",
      "mRNA = UGGUAUUGG\n",
      "UGG = 2\n",
      "UAU = 1\n",
      "\n",
      "mRNA = UGGUACUGG\n",
      "UGG = 2\n",
      "UAC = 1\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Provide the frequency of each RNA codon encoding a given aminoacid, in a DNA \n",
    "sequence\n",
    "PS: \n",
    "• Please provide all possibilities of the codon \n",
    "combination.\n",
    "• Input max. 3 Aminoacids.\n",
    "\n",
    "'''\n",
    "\n",
    "import itertools\n",
    "\n",
    "# Codon table for amino acids to codons\n",
    "amino_acid_to_codons = {\n",
    "    'A': ['GCU', 'GCC', 'GCA', 'GCG'],  # Alanine\n",
    "    'C': ['UGU', 'UGC'],                # Cysteine\n",
    "    'D': ['GAU', 'GAC'],                # Aspartic Acid\n",
    "    'E': ['GAA', 'GAG'],                # Glutamic Acid\n",
    "    'F': ['UUU', 'UUC'],                # Phenylalanine\n",
    "    'G': ['GGU', 'GGC', 'GGA', 'GGG'],  # Glycine\n",
    "    'H': ['CAU', 'CAC'],                # Histidine\n",
    "    'I': ['AUU', 'AUC', 'AUA'],         # Isoleucine\n",
    "    'K': ['AAA', 'AAG'],                # Lysine\n",
    "    'L': ['UUA', 'UUG', 'CUU', 'CUC', 'CUA', 'CUG'],  # Leucine\n",
    "    'M': ['AUG'],                       # Methionine\n",
    "    'N': ['AAU', 'AAC'],                # Asparagine\n",
    "    'P': ['CCU', 'CCC', 'CCA', 'CCG'],  # Proline\n",
    "    'Q': ['CAA', 'CAG'],                # Glutamine\n",
    "    'R': ['CGU', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],  # Arginine\n",
    "    'S': ['UCU', 'UCC', 'UCA', 'UCG', 'AGU', 'AGC'],  # Serine\n",
    "    'T': ['ACU', 'ACC', 'ACA', 'ACG'],  # Threonine\n",
    "    'V': ['GUU', 'GUC', 'GUA', 'GUG'],  # Valine\n",
    "    'W': ['UGG'],                       # Tryptophan\n",
    "    'Y': ['UAU', 'UAC'],                # Tyrosine\n",
    "    '*': ['UAA', 'UAG', 'UGA'],         # Stop codons\n",
    "}\n",
    "\n",
    "# Function to generate all possible mRNA sequences from an amino acid sequence\n",
    "def codon_combinations(amino_acid_sequence):\n",
    "    codon_list = [amino_acid_to_codons[aa] for aa in amino_acid_sequence if aa in amino_acid_to_codons]\n",
    "    \n",
    "    # Generate all possible combinations of codons using Cartesian product\n",
    "    possible_mRNA_sequences = [''.join(codon_set) for codon_set in itertools.product(*codon_list)]\n",
    "    \n",
    "    return possible_mRNA_sequences\n",
    "\n",
    "# Function to calculate codon frequency in a given mRNA sequence\n",
    "def codon_frequency(mRNA_sequence):\n",
    "    codon_count = {}\n",
    "    for i in range(0, len(mRNA_sequence), 3):\n",
    "        codon = mRNA_sequence[i:i+3]\n",
    "        if codon in codon_count:\n",
    "            codon_count[codon] += 1\n",
    "        else:\n",
    "            codon_count[codon] = 1\n",
    "    return codon_count\n",
    "\n",
    "# Function to print codon combinations and frequencies from amino acid sequence\n",
    "def codon_frequency_from_amino_acids(amino_acid_sequence):\n",
    "    # Generate all possible mRNA sequences\n",
    "    possible_mRNA_sequences = codon_combinations(amino_acid_sequence)\n",
    "    \n",
    "    print(f\"Input Aminoacid = {amino_acid_sequence}\")\n",
    "    \n",
    "    for mRNA_sequence in possible_mRNA_sequences:\n",
    "        print(f\"\\nmRNA = {mRNA_sequence}\")\n",
    "        codon_count = codon_frequency(mRNA_sequence)\n",
    "        for codon, count in codon_count.items():\n",
    "            print(f\"{codon} = {count}\")\n",
    "\n",
    "# Example usage:\n",
    "amino_acid_sequence = input(\"Enter amino acids (max 3 letters): \").upper()\n",
    "if len(amino_acid_sequence) <= 3:\n",
    "    codon_frequency_from_amino_acids(amino_acid_sequence)\n",
    "else:\n",
    "    print(\"Invalid input. Please enter up to 3 amino acids.\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
